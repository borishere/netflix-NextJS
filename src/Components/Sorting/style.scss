.sorting {
  .sorting-title {
    margin-right: 35px;
    font-size: 16px;
    color: #ffffff;
    mix-blend-mode: normal;
    opacity: 0.6;
  }

  .sorting-items {
    font-weight: 500;
    font-size: 16px;
    line-height: 20px;
    background: rgba(0, 0, 0, 0);
    color: #fff;
    border: 0;
    text-transform: uppercase;
    cursor: pointer;
  }
}
