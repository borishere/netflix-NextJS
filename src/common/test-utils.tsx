import React, { FC, ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ModalContext } from '../Context/ModalContext';
import { Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../app/store';
import { createMemoryHistory } from 'history';
import { moviesApi } from '../services/movies';
import { setupListeners } from '@reduxjs/toolkit/dist/query';
import moviesReducer from '../app/moviesSlice';
import { configureStore } from '@reduxjs/toolkit';

const history = createMemoryHistory();


const AllTheProviders: FC = ({ children }) => {
  return (
    <Router location='/' navigator={history}>
      <Provider store={store}>
        <ModalContext.Provider value={{}}>
          {children}
        </ModalContext.Provider>
      </Provider>
    </Router>
  );
};

function renderWithStore(
  ui: ReactElement,
  {
    store = configureStore({
      reducer: {
        [moviesApi.reducerPath]: moviesApi.reducer,
        movies: moviesReducer

      },
      middleware: (getDefaultMiddleware: any) =>
        getDefaultMiddleware().concat(moviesApi.middleware)
    }),
    ...renderOptions
  } = {}
) {
  setupListeners(store.dispatch);

  function Wrapper({ children }: any) {
    const history = createMemoryHistory();

    return (
      <Provider store={store}>
        <Router location='/' navigator={history}>
          {children}
        </Router>
      </Provider>
    );
  }

  store.dispatch(moviesApi.util.resetApiState());

  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

const root = document.createElement('div');
root.id = 'root';

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, container: document.body.appendChild(root), ...options });

export * from '@testing-library/react';
export { customRender, renderWithStore };