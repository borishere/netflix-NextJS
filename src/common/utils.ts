export function addParamToExistsSearchParams(key: string, value: string): URLSearchParams {
  const params = new URLSearchParams(window.location.search);

  params.set(key, value);

  return params;
}

export function deleteParamFromExistsSearchParams(value: string): URLSearchParams {
  const params = new URLSearchParams(window.location.search);

  params.delete(value);

  return params;
}

export const BASE_API_URL: string = 'http://localhost:4000/';