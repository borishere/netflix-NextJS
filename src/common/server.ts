import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { BASE_API_URL } from './utils';
import mockData from './__mocks__/movies.json';

let movies: any[] = Array.from(mockData);

export const server = setupServer(
  rest.get(`${BASE_API_URL}movies`, (req, res, ctx) => {
    return res(ctx.json(movies.slice(0, 2)));
  }),

  rest.get(`${BASE_API_URL}movies:id`, (req, res, ctx) => {
    const { id } = req.params;
    const movie = movies.find((movie) => movie.id === id);

    return res(ctx.json(movie));
  }),

  rest.post(`${BASE_API_URL}movies`, (req, res, ctx) => {
    movies.push(req.body);

    return res(ctx.status(200));
  }),

  rest.put(`${BASE_API_URL}movies`, (req, res, ctx) => {
    const requestMovie: any = req.body;

    const movieIndex = movies.findIndex((m) => m.id === requestMovie.id);

    if (movieIndex < 0) {
      return res(ctx.status(404));
    }

    movies[movieIndex] = requestMovie;

    return res(ctx.status(200));
  }),

  rest.delete(`${BASE_API_URL}movies/:id`, (req, res, ctx) => {
    const { id } = req.params;

    movies = movies.filter((m) => m.id !== id);

    return res(ctx.status(200));
  })
);
