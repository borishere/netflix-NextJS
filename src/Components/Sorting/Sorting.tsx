import React, { FC, useEffect, useState } from 'react';
import { useSearchParams, useNavigate } from 'react-router-dom';
import { initialState } from '../../app/moviesSlice';
import { useAppDispatch } from '../../Hooks/hooks';
import { setSortBy } from '../../app/moviesSlice';
import { addParamToExistsSearchParams } from '../../common/utils';
import './style.scss';

export const Sorting: FC = () => {
  const [value, setValue] = useState<string>(initialState.sortBy);
  const [searchParams] = useSearchParams();
  const sortBy = searchParams.get('sortBy');
  const dispatch = useAppDispatch();
  const navigate = useNavigate();

  const onSort = (e: React.ChangeEvent<HTMLSelectElement>) => {
    const value = e.target.value;
    setValue(value);

    const params = addParamToExistsSearchParams('sortBy', value);

    navigate({search: params.toString()});
  };

  useEffect(() => {
    dispatch(setSortBy(value));
  }, [value, dispatch]);

  useEffect(() => {
    if (sortBy) {
      setValue(sortBy);
    }
  }, [sortBy]);


  return (
    <div className='sorting'>
      <span className='sorting-title'>SORT BY</span>

      <select
        className='sorting-items'
        value={value}
        onChange={onSort}
        aria-label='sorting-select'
      >
        <option value='release_date'>release date</option>
        <option value='vote_count'>rating</option>
        <option value='title'>title</option>
      </select>
    </div>
  );
};
