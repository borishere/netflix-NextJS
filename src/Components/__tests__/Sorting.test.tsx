import React from 'react';
import { screen, customRender } from '../../common/test-utils';
import { Sorting } from '../Sorting/Sorting';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

describe('Sorting', () => {
  test('Renders', () => {
    const { asFragment } = customRender(<Sorting />);

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText('SORT BY')).toBeInTheDocument();
  });

  test('Renders options', () => {
    customRender(<Sorting />);

    const sortingSelect = screen.getByLabelText('sorting-select');

    userEvent.selectOptions(sortingSelect, 'vote_count');

    expect(sortingSelect).toHaveTextContent('rating');
  });
});