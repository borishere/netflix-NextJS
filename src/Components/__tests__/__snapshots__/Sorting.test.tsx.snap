// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sorting Renders 1`] = `
<DocumentFragment>
  <div
    class="sorting"
  >
    <span
      class="sorting-title"
    >
      SORT BY
    </span>
    <select
      aria-label="sorting-select"
      class="sorting-items"
    >
      <option
        value="release_date"
      >
        release date
      </option>
      <option
        value="vote_count"
      >
        rating
      </option>
      <option
        value="title"
      >
        title
      </option>
    </select>
  </div>
</DocumentFragment>
`;
